src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -0,0 +1,382 @@
+        {
+            lock (_sync)
+            {
+                if (_currentFlushTcs != null)
+                {
+                    return new ValueTask<FlushResult>(_currentFlushTcs.Task);
+                }
+
+                if (_bytesBuffered > 0)
+                {
+                    CopyAndReturnSegmentsUnsynchronized();
+                }
+
+                var flushTask = _innerPipeWriter.FlushAsync(cancellationToken);
+
+                if (flushTask.IsCompletedSuccessfully)
+                {
+                    if (_currentFlushTcs != null)
+                    {
+                        CompleteFlushUnsynchronized(flushTask.GetAwaiter().GetResult(), null);
+                    }
+
+                    return flushTask;
+                }
+
+                // Use a TCS instead of something resettable so it can be awaited by multiple awaiters.
+                _currentFlushTcs = new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+                return FlushAsyncAwaited(flushTask, cancellationToken);
+            }
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncAwaited(ValueTask<FlushResult> flushTask, CancellationToken cancellationToken)
+        {
+            try
+            {
+                // This while (true) does look scary, but the real continuation condition is at the start of the loop
+                // after the await, so the _sync lock can be acquired.
+                while (true)
+                {
+                    var flushResult = await flushTask;
+
+                    lock (_sync)
+                    {
+                        if (_bytesBuffered == 0)
+                        {
+                            CompleteFlushUnsynchronized(flushResult, null);
+                            return flushResult;
+                        }
+
+                        CopyAndReturnSegmentsUnsynchronized();
+
+                        flushTask = _innerPipeWriter.FlushAsync(cancellationToken);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                lock (_sync)
+                {
+                    CompleteFlushUnsynchronized(default, ex);
+
+                    // Ensure the exception still gets observed by the original caller who started the flush loop
+                    // And is observing this method directly instead of through the TCS.
+                    throw;
+                }
+            }
+        }
+
+        // This is not exposed to end users. Throw so we find out if we ever start calling this.
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Abort()
+        {
+            lock (_sync)
+            {
+                _aborted = true;
```

Comment:
Can this be complete rather than Abort?

Reply:
I like Abort() because it behaves differently than PipeWriter.Complete() normally does.

It doesn't actually call Complete() on the underlying pipe, so it doesn't communicate completion of the output. The inner pipe completion instead happens later in the project bedrock pipeline. Abort() also doesn't care about any exceptions, it could be called mid-flush, and it doesn't actually cause future API calls to fail. Since it's semantically different than PipeWriter.Complete(), I felt it was important to make it another method.

Reply:
Sounds good, I just realized that we store an instance of the ConcurrentPipeWriter rather than the PipeWriter so we can have extra methods on it.